### PR TITLE
Add evaluation metrics for multilabel task

### PR DESCRIPTION
**Pull request by LXXXXR**
Add mAP, CP, CR, CF1, OP, OR, OF1 as multi-label task evaluation metrics:
1. Unittests are put in multilabel_eval_metrics.py temporarily, and I will put them in test_dataset.py once multi-label datasets are ready.
2. In papers, some report topk performance while others use threshold, so both methods are implemented for future convenience.
3. For mAP, the method provided by PASCAL VOC after 2010 is used.

### CODE REVIEW

File: mmcls/core/evaluation/multilabel_eval_metrics.py [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+import torch
+
+
+def average_performance(pred, target, thrs=None, k=None):
+    """Calculate CP, CR, CF1, OP, OR, OF1, where C stands for per-class
+        average, O stands for overall average, P stands for precision, R
+        stands for recall and F1 stands for F1-score
+
+    Args:
+        pred (torch.Tensor | np.ndarray): The model prediction.
```
Comment (ZwwWayne) [inline mmcls/core/evaluation/multilabel_eval_metrics.py:13]: Indicate the shape of pred and target

File: .github/workflows/build.yml [hunk 2]
```diff
@@ -54,6 +54,10 @@ jobs:
         if: ${{matrix.torchvision < 0.5}}
       - name: Install PyTorch
         run: pip install --use-deprecated=legacy-resolver torch==${{matrix.torch}} torchvision==${{matrix.torchvision}} -f https://download.pytorch.org/whl/torch_stable.html
+      # temporary solution until new version of mmcv is avaliable
```
Comment (hellock) [inline .github/workflows/build.yml:57]: The new version has been published.

File: mmcls/core/evaluation/mean_ap.py [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+
+def average_precision(pred, target):
+    """ Calculate the average precision for a single class
+
+    Args:
+        pred (np.ndarray): The model prediction with shape (N, C), where C is
+            the number of classes.
+        target (np.ndarray): The target of each prediction with shape (N, C),
+            where C is the number of classes.
+
```
Comment (ycxioooong) [inline mmcls/core/evaluation/mean_ap.py:13]: target: N x C or N x 1?

File: mmcls/core/evaluation/mean_ap.py [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+    Returns:
+        float: a single float as average precision value.
+    """
+    eps = np.finfo(np.float32).eps
```
Comment (ycxioooong) [inline mmcls/core/evaluation/mean_ap.py:17]: there is a more accurate way to avoid division by zero

File: mmcls/core/evaluation/mean_ap.py [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+    sort_inds = np.argsort(-pred)
+    sort_target = target[sort_inds]
+
+    # count true positive examples
+    p_inds = sort_target == 1
+    tp = np.cumsum(p_inds)
+    total_p = tp[-1]
+
+    # count not difficult examples
+    pn_inds = sort_target != -1
```
Comment (ycxioooong) [inline mmcls/core/evaluation/mean_ap.py:29]: add an option for ignoring difficult samples, instead of hard code.

File: mmcls/core/evaluation/mean_ap.py [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+    pn = np.cumsum(pn_inds)
+
+    tp[np.logical_not(p_inds)] = 0
+    precision = tp / (pn + eps)
```
Comment (ycxioooong) [inline mmcls/core/evaluation/mean_ap.py:33]: also, use a more elegant way to replace eps

File: mmcls/core/evaluation/multilabel_eval_metrics.py [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+        tuple: (CP, CR, CF1, OP, OR, OF1)
+    """
+    if isinstance(pred, torch.Tensor) and isinstance(target, torch.Tensor):
+        pred = pred.numpy()
+        target = target.numpy()
+    elif not (isinstance(pred, np.ndarray) and isinstance(target, np.ndarray)):
+        raise TypeError('pred and target should both be torch.Tensor or'
+                        'np.ndarray')
+    if thrs is None and k is None:
+        thrs = 0.5
```
Comment (ycxioooong) [inline mmcls/core/evaluation/multilabel_eval_metrics.py:33]: check the default value according to Head implementations.

File: mmcls/core/evaluation/multilabel_eval_metrics.py [hunk 8]
```diff
@@ -0,0 +1,68 @@
...
+        stands for recall and F1 stands for F1-score
+
+    Args:
+        pred (torch.Tensor | np.ndarray): The model prediction with shape
+            (N, C), where C is the number of classes.
+        target (torch.Tensor | np.ndarray): The target of each prediction with
+            shape (N, C), where C is the number of classes. 1 stands for
+            positive examples, 0 stands for negative examples and -1 stands for
+            difficult examples.
+        thrs (float): The confidence threshold. Defaults to None.
```
Comment (ycxioooong) [inline mmcls/core/evaluation/multilabel_eval_metrics.py:19]: thrs ==> thr

File: mmcls/core/evaluation/multilabel_eval_metrics.py [hunk 9]
```diff
@@ -0,0 +1,68 @@
...
+        thrs = 0.5
+        warnings.warn('Neither thrs nor k is given, set thrs as 0.5 by '
+                      'default.')
+    elif thrs is not None and k is not None:
+        warnings.warn('Both thrs and k are given, use threshold in favor of '
+                      'top-k.')
+
+    assert pred.shape == target.shape
+
+    eps = np.finfo(np.float32).eps
```
Comment (ycxioooong) [inline mmcls/core/evaluation/multilabel_eval_metrics.py:42]: replace eps here

File: mmcls/core/evaluation/multilabel_eval_metrics.py [hunk 10]
```diff
@@ -0,0 +1,68 @@
...
+        # a label is predicted positive if the confidence is no lower than thrs
+        p_inds = pred >= thrs
+
+    else:
+        # top-k labels will be predicted positive for any example
+        sort_inds = np.argsort(-pred, axis=1)
+        sort_inds_ = sort_inds[:, :k]
+        inds = np.indices(sort_inds_.shape)
+        p_inds = np.zeros_like(pred)
+        p_inds[inds[0], sort_inds_] = 1
```
Comment (ycxioooong) [inline mmcls/core/evaluation/multilabel_eval_metrics.py:54]: np.arange

File: mmcls/core/evaluation/multilabel_eval_metrics.py [hunk 11]
```diff
@@ -0,0 +1,68 @@
...
+
+    tp = (p_inds * target) == 1
+    fp = (p_inds * (1 - target)) == 1
+    fn = ((1 - p_inds) * target) == 1
+
+    precision_class = tp.sum(0) / (tp.sum(0) + fp.sum(0) + eps)
```
Comment (ycxioooong) [inline mmcls/core/evaluation/multilabel_eval_metrics.py:60]: use axis=0 for readability

**Review by ycxioooong — CHANGES_REQUESTED** [review]

File: mmcls/core/evaluation/mean_ap.py [hunk 12]
```diff
@@ -0,0 +1,63 @@
...
+import torch
+
+
+def average_precision(pred, target):
+    """ Calculate the average precision for a single class
+
+    Args:
+        pred (np.ndarray): The model prediction with shape (N, ).
+        target (np.ndarray): The target of each prediction with shape (N, ).
+
```
Comment (MeowZheng) [inline mmcls/core/evaluation/mean_ap.py:11]: it would help to better understand by writing evaluation formula in docstring.
 
FYI. https://github.com/scikit-learn/scikit-learn/blob/42aff4e2e/sklearn/metrics/_ranking.py#L118
https://scikit-learn.org/stable/modules/generated/sklearn.metrics.average_precision_score.html?highlight=average_precision#sklearn.metrics.average_precision_score

File: mmcls/core/evaluation/mean_ap.py [hunk 13]
```diff
@@ -0,0 +1,63 @@
...
+        float: a single float as average precision value.
+    """
+    eps = np.finfo(np.float32).eps
+
+    # sort examples
+    sort_inds = np.argsort(-pred)
+    sort_target = target[sort_inds]
+
+    # count true positive examples
+    p_inds = sort_target == 1
```
Comment (MeowZheng) [inline mmcls/core/evaluation/mean_ap.py:22]: Is "p" short for "postive"? If it is, "pos" might be better than "p".

File: mmcls/core/evaluation/multilabel_eval_metrics.py (hunk 7 above)
Comment (LXXXXR) [reply mmcls/core/evaluation/multilabel_eval_metrics.py]: I will put sigmoid in simple test in Head implementation, therefore, perhaps we can leave default thr as 0.5 here?

File: mmcls/core/evaluation/mean_ap.py (hunk 5 above)
Comment (LXXXXR) [reply mmcls/core/evaluation/mean_ap.py:36]: Could you please be more specific about this comment? The code now can calculate map whether there are difficult examples. Thanks.

File: mmcls/core/evaluation/multilabel_eval_metrics.py (hunk 10 above)
Comment (LXXXXR) [reply mmcls/core/evaluation/multilabel_eval_metrics.py:54]: I am not sure how I should use it here. Could you kindly specify it a little bit? Are you suggesting use np.arrange and stack them together?

File: mmcls/core/evaluation/multilabel_eval_metrics.py [hunk 14]
```diff
@@ -0,0 +1,70 @@
...
+        # a label is predicted positive if the confidence is no lower than thr
+        p_inds = pred >= thr
+
+    else:
+        # top-k labels will be predicted positive for any example
+        sort_inds = np.argsort(-pred, axis=1)
+        sort_inds_ = sort_inds[:, :k]
+        inds = np.indices(sort_inds_.shape)
+        p_inds = np.zeros_like(pred)
+        p_inds[inds[0], sort_inds_] = 1
```
Comment (MeowZheng) [inline mmcls/core/evaluation/multilabel_eval_metrics.py:54]: if p is short for "positive", "pos" may better than "p"

File: mmcls/core/evaluation/multilabel_eval_metrics.py [hunk 15]
```diff
@@ -0,0 +1,70 @@
...
+                        'np.ndarray')
+    if thr is None and k is None:
+        thr = 0.5
+        warnings.warn('Neither thr nor k is given, set thr as 0.5 by '
+                      'default.')
+    elif thr is not None and k is not None:
+        warnings.warn('Both thr and k are given, use threshold in favor of '
+                      'top-k.')
+
+    assert pred.shape == target.shape
```
Comment (MeowZheng) [inline mmcls/core/evaluation/multilabel_eval_metrics.py:40]: please give some comments for this assert

File: mmcls/core/evaluation/mean_ap.py [hunk 16]
```diff
@@ -0,0 +1,72 @@
...
+        float: A single float as mAP value.
+    """
+    if isinstance(pred, torch.Tensor) and isinstance(target, torch.Tensor):
+        pred = pred.numpy()
+        target = target.numpy()
+    elif not (isinstance(pred, np.ndarray) and isinstance(target, np.ndarray)):
+        raise TypeError('pred and target should both be torch.Tensor or'
+                        'np.ndarray')
+
+    assert pred.shape == target.shape
```
Comment (MeowZheng) [inline mmcls/core/evaluation/mean_ap.py:66]: please add some comments

**Review by ycxioooong — APPROVED** [review]
LGTM now
